src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.PieceDirectory.cs

```diff
@@ -0,0 +1,456 @@
+        //  At any point in time, the members of the sorted array _pieceStreamInfoArray are *adjacent* pieces whose
+        //  index reflects their order in the part. On the other hand, _pieceStreamInfoArray is not required to represent
+        //  all the pieces in the part. (The last useful entry in _pieceStreamInfoArray is at _indexOfLastPieceStreamInfoAccessed.)
+        //  At init time, the first piece descriptor is loaded into _pieceStreamInfoArray. Thereafter, descriptors are loaded
+        //  on demand in piece order and without any gap between them.
+
+        private sealed class PieceDirectory
+        {
+            /// <summary>
+            /// PieceStreamInfo Descriptor.
+            /// Provides access to piece's stream and start offset.
+            /// </summary>
+            /// <remarks>
+            /// This is a private class. All of the above information, together with position
+            /// information (piece number, last-piece status), is accessed through the
+            /// PieceDirectory class by providing offsets or piece numbers.
+            /// </remarks>
+            private sealed class PieceStreamInfo : IComparable<PieceStreamInfo>
+            {
+
+                /// <summary>
+                /// Initialize a PieceStreamInfo.
+                /// </summary>
+                internal PieceStreamInfo(Stream stream, long pieceStart)
+                {
+                    Debug.Assert(stream != null);
+                    Debug.Assert(pieceStart >= 0);
+                    Stream = stream;
+                    StartOffset = pieceStart;
+                }
+
+                /// <summary>
+                /// Returns the startOffset of the piece in the part
+                /// </summary>
+                internal long StartOffset { get; }
+
+
+                /// <summary>
+                /// Returns the stream for this piece
+                /// </summary>
+                internal Stream Stream { get; set; }
+
+
+                /// <summary>
+                /// IComparable.CompareTo implementation which allows sorting
+                /// descriptors by range of offsets.
+                /// </summary>
+                int IComparable<PieceStreamInfo>.CompareTo(PieceStreamInfo? pieceStreamInfo)
+                {
+                    return pieceStreamInfo == null
+                        ? 1
+                        : StartOffset.CompareTo(pieceStreamInfo.StartOffset);
+                }
+
+            }
+
+            /// <summary>
+            /// Load the descriptor of the first piece into _pieceStreamInfoArray and store context data
+            /// extracted from the arguments.
+            /// </summary>
+            internal PieceDirectory(List<ZipPackagePartPiece> sortedPieceInfoList, ZipStreamManager zipStreamManager, FileAccess access)
+            {
+                if (sortedPieceInfoList == null)
+                    throw new ArgumentNullException(nameof(sortedPieceInfoList));
```

Comment:
For non-public methods, we usually do not check for null and let the caller ensure it instead.